### ISSUE TITLE
nonintuitive counts after selecting facet value

### ISSUE THREAD

**Issue by andrewsu** [NIAID-Data-Ecosystem/nde-portal#145]
If I start at https://data-staging.niaid.nih.gov/portal/search/?q= with no search specified, I see this section of the facet controls for "Sources":

![image](https://user-images.githubusercontent.com/2635409/208772996-0aa1c41e-3e80-40e0-a89a-bd0f74392668.png)

If I select one option (e.g., "zenodo"), I see this:

![image](https://user-images.githubusercontent.com/2635409/208773137-b20a57ad-aa64-42f5-a8d9-341d671d4268.png)

The counts for the other sources has been set to zero.  I understand why the facet counts update in the context of _other_ sections of the faceting controls (we want to show the user the number of records that would result when adding another filter by "AND"), but since this is "OR"ed, the counts aren't quite right.  But then what is the appropriate count to show when you have facets specified in multiple sections?  I think the current behavior is non-intuitive, but it may be that this is no completely intuitive logic possible.  

Bottom line, it might be that we are already at the optimal solution, but just creating the ticket at least to track that we've discussed it...

**Comment by gtsueng** [comment]
Agree it seems a bit on the strange side, considering that some of the other filters do behave within expectation. 
Host filter example:
![image](https://github.com/NIAID-Data-Ecosystem/nde-portal/assets/9200138/eb0eea65-b945-4f58-86b2-fa7802ca00c0)
But the reverse is not as intuitive:
![image](https://github.com/NIAID-Data-Ecosystem/nde-portal/assets/9200138/d63aba6d-d5ee-476a-b254-783eb2940a8b)

Given that some filters behave as how @andrew would expect, but others do not, it seems like it's fixable.

Filters that behave as expected:
`Health condition`
`Host species`
`Pathogen species`
`Funding`
-- These all appear to be fields we have standardized. Is it because they can be mapped in the ES as keywords (allowing for easy aggregation?)

Filters that do not behave as expected:
`Repository` <-- should also be standardized since it will be set by our parsers
`Conditions of access` <-- should be standardized (one of four enum values)
`Variable measured`
`Measurement Technique`

**Comment by gtsueng** [comment]
@candicecz @EmilyHaag Do either of you remember if this behavior was intended based on input from the NIAID team?

**Comment by candicecz** [comment]
The filters currently all have the same functionality including the filters for  `Health condition`, `Host species`,`Pathogen species`, `Funding`. Many of the values have datasets in common so it's only once you scroll far enough that you see values with a zero count.
![Screenshot from 2024-02-07 16-07-14](https://github.com/NIAID-Data-Ecosystem/nde-portal/assets/25373313/35e8ced7-ac73-49e4-867c-ce2c87ac66bf)

The suggested functionality per this discussion seems to be to maintain the counts when filters are toggled **within** a facet and only update counts when other facets are changed. Similar to ebay's filter functionality : https://www.ebay.com/sch/i.html?_from=R40&_nkw=dog+toys&_sacat=0&rt=nc&Dog%2520Size=S&_dcat=177791

This feature is complicated to implement and requires some consensus on the desired behaviour before moving forward. We will re-visit this issue pending a higher priority status.

**Comment by gtsueng** [comment]
Thanks for looking into that!

**Comment by andrewsu** [comment]
I see the issue now of how there are multiple possible behaviors that need discussion/consensus. If I search "asthma", I currently see 8954 results.  If I look at facets for "Pathogen species", here's what I see:

No facet selected: 8954
[Enterovirus](https://data.niaid.nih.gov/search?q=%22Asthma%22&from=1&filters=%28infectiousAgent.displayName.raw%3A%28%22Enterovirus%22%29%29) selected: 25 
[Human rhinovirus sp](https://data.niaid.nih.gov/search?q=%22Asthma%22&from=1&filters=%28infectiousAgent.displayName.raw%3A%28%22Human+rhinovirus+sp.%22%29%29) selected: 8
[Enterovirus and Human rhinovirus sp](https://data.niaid.nih.gov/search?q=%22Asthma%22&from=1&filters=%28infectiousAgent.displayName.raw%3A%28%22Enterovirus%22+OR+%22Human+rhinovirus+sp.%22%29%29) selected: 27

So from a UI perspective, if I have "Enterovirus" selected, does the number on "Human rhinovirus sp" say 8 (because that's how many datasets match that value on its own) or does it say 2 (because that's how many additional datasets would be added if it would be selected)?

... and unfortunately ebay doesn't turn out to be a good model here, since they're a little imprecise with their facet counts.  For example, "[green dog toys for aggressive chewers](https://www.ebay.com/sch/i.html?_nkw=green+dog+toys+for+aggressive+chewers)" currently has 844 results. The Type facet for "chew toy" shows 460, but selecting it shows [443 results](https://www.ebay.com/sch/i.html?_nkw=green+dog+toys+for+aggressive+chewers&rt=nc&Type=Chew%2520Toy&_dcat=177791). Combining it with Type facet for "Ball" does not result in an intuitive count.

**Comment by gtsueng** [comment]
Some potential next steps:
1. Systematically review the filter count behavior from the following websites to identify types of filter behavior:
    - [ ] big online retailers: Vons, Walmart, Costco, Carmax, Ebay 
    - [ ] Well-used databases: [ClinicalTrials.gov](http://clinicaltrials.gov/), [Zenodo.org](http://zenodo.org/) 
2. Conduct user studies to determine:
    - [ ] Are filter counts valuable for the site? The value of filter counts seem to diminish above 1000 and are omitted altogether in these sites: Pubmed, Amazon, Target, google dataset search, Apple
    - [ ] If they are valuable, what is the expected behavior? More like an online retailer? More like a database?

**Comment by gtsueng** [comment]
Hi @lisa-mml,

As I understand it, the filters currently all have the same functionality including the filters for Health condition, Host species,Pathogen species, Funding. Changing the filter behavior is complicated and we think it would be better to identify the desired behavior before making any changes and potentially worsening the problem. It's currently unclear what counts to display when we don't zero/grey out unselected filter options.

For example, if I select OMICs-DI, should the counts of unselected GEO records be all of GEO or just the records in GEO that overlap with OMICs-DI? Some users prefer the former, while others prefer the latter--we have yet to nail down the best behavior.

If this is a priority, we can start with doing research on identifying the best solution; while our front-end team works on disease pages, program collection pages, etc -- that have previously been prioritized over this.

**Comment by lisa-mml** [comment]
Thanks @gtsueng. I'll bring this to our next internal discussion and get back to you with any updates.

**Comment by gtsueng** [comment]
Systematic review of filter count behavior: https://docs.google.com/spreadsheets/d/1QzvlNY8yeMWPDkN6VlGJU4PBBTqeFp2-aK_H3MSBacY/edit?gid=1713308086#gid=1713308086

Key takeaways:
* most sites keep the counts even when a filter option is selected
* most sites hide filter options when the count drops to zero
* counts on retailer sites are often ballpark counts and will change between what's displayed when a filter is/is not selected
* counts for research resources do not fluctuate to the extent that retailer sites do (and in the case of Zenodo, the filter option counts don't change at all).

**Comment by EmilyHaag** [comment]
Results of internal study with Scripps team members: https://docs.google.com/presentation/d/1EpUkvVfb-5My4VqTkN-QjpsKaGKCaEf3/edit?usp=sharing&ouid=114065631315523104759&rtpof=true&sd=true

**Comment by andrewsu** [comment]
(This is obviously a Claude-authored comment, but seems to check out...)

Re-tested on production today (2026-09-14). Still reproducible, and in one respect worse than the 2022 report: the unselected values in the same facet are no longer *zeroed*, they are **removed from the list entirely, effectively preventing multiselection**.

**Sources, no search term.** Before: `NCBI GEO 8,826,454` / `Figshare 2,177,036` / `Zenodo 649,586` / … After clicking `Zenodo`, the whole Sources facet is:

```
IID Repositories
  Project Tycho                597
Generalist Repositories
  Zenodo                   649,586
  Dryad Digital Repository  58,896
```

NCBI GEO and Figshare are gone, and the in-facet search box can't find them either (typing "GEO" or "Figshare" returns nothing). So `Zenodo OR NCBI GEO` — 9,476,383 records, which the API answers fine — is **unreachable by clicking**, even though the Knowledge Center [tells users](https://data.niaid.nih.gov/knowledge-center/filtering-results) that "Selecting within a group looks for **any** of those filters." In practice a single-valued facet like Sources behaves as single-select once you pick a value.

**Pathogen Species, `q=asthma`.** Selecting `Enterovirus` takes `Respiratory syncytial virus` from 60 → 17 and `Human rhinovirus sp.` to 5, and drops `Unspecified (11,480)` off the list. The surviving numbers are *intersections* — "records that have Enterovirus **and** RSV" — which is the one reading of that number nobody wants: it isn't the value's own count, and it isn't the marginal records you'd add.

Worth noting for @lisa-mml's request to "avoid zeroing/greying out unselected filter options": this isn't a display decision that can be tweaked in the UI. An Elasticsearch `terms` aggregation never emits a zero-count bucket, so those options are absent from the API response, not hidden by the frontend. Fixing the display requires fixing the aggregation.

---

### There is a standard design for this, and it has a name

"Multi-select faceting" / drill-down exclusion. The rule is one line:

> **When computing a facet's own counts, ignore that facet's selections. Every *other* facet's selections still apply.**

Result counts stay fully filtered; only the facet counts change. This is a first-class feature in both major engines, and both document it with exactly this issue as the motivating example:

- Elasticsearch — [`post_filter`](https://www.elastic.co/guide/en/elasticsearch/reference/current/filter-search-results.html): "the search hits are filtered after the aggregations are calculated." The worked example in the docs is a shirt shop where the user has picked red and you still want to "tell the user how many Gucci shirts are available in **other colors**."
- Solr — [Tagging and Excluding Filters](https://solr.apache.org/guide/solr/latest/query-guide/faceting.html): `fq={!tag=dt}doctype:pdf&facet.field={!ex=dt}doctype`.

This also matches what @gtsueng's [systematic review](https://docs.google.com/spreadsheets/d/1QzvlNY8yeMWPDkN6VlGJU4PBBTqeFp2-aK_H3MSBacY/edit?gid=1713308086#gid=1713308086) found to be the *common* behavior across the sites surveyed: "No change in filter counts within a filter even after the selection of an option in that filter."

And it answers the question I asked in [my 2025-03-17 comment](https://github.com/NIAID-Data-Ecosystem/nde-portal/issues/145#issuecomment-2730130897) — should "Human rhinovirus sp." show 8 or 2? The standard answer is **8**: its count in the rest of the query, ignoring the Pathogen Species selection. The marginal-addition reading (2) is order-dependent — it changes depending on which value you clicked first — and no site in the review uses it. 8 is stable, and adding it to the selection lands you on the 27 the facet arithmetic implies.

### The part I didn't expect: our API already supports it

`post_filter` is a documented parameter on the NDE query API (it's in `/v1/spec`), and it has the semantics we need.

Same query, two ways — both are plain GETs, so you can just click them:

**1. Today's portal behavior** — the selection goes in `extra_filter`: [`…&extra_filter=infectiousAgent.displayName.raw:("Enterovirus")`](https://api.data.niaid.nih.gov/v1/query?q=asthma&size=0&facets=infectiousAgent.displayName.raw&facet_size=6&extra_filter=infectiousAgent.displayName.raw%3A%28%22Enterovirus%22%29)

> `total: 109` — facet reads Enterovirus 109, RSV 17, Human RSV 15, … (intersections)

**2. Proposed** — the same selection goes in `post_filter`: [`…&post_filter=infectiousAgent.displayName.raw:("Enterovirus")`](https://api.data.niaid.nih.gov/v1/query?q=asthma&size=0&facets=infectiousAgent.displayName.raw&facet_size=6&post_filter=infectiousAgent.displayName.raw%3A%28%22Enterovirus%22%29)

> `total: 109` — facet reads SARS-CoV-2 115, Enterovirus 109, RSV 61, E. coli 53, …

Same correctly-filtered total of 109, but the facet keeps its pre-selection counts — bit-for-bit identical to the unfiltered aggregation. No backend change required.

One caveat I verified: `post_filter` is global to the request, so it's also excluded from *every other* facet's aggregation in that same response. So the correct wiring is **1 + N requests**, where N = the number of facets that currently have a selection:

- **Base request** — all selections in `extra_filter`. Drives the result total and every facet with no selections.
- **One request per selected facet F** — F's selections in `post_filter`, all other facets' selections in `extra_filter`. Drives F's counts.

N is typically 1–3, and [`useFilterQueries`](https://github.com/NIAID-Data-Ecosystem/nde-portal/blob/main/src/views/search/components/filters/hooks/useFilterQueries.ts) already fans out four parallel scoped aggregations and routes each filter config to one of them, so this is additive to an existing pattern rather than an architectural change. The single place that currently causes the behavior is that every filter's counts come from one aggregation whose `extra_filter` includes that filter's own selections.

Two useful side effects: values stop vanishing from a facet you've selected in (so multi-select becomes reachable, addressing @lisa-mml's point 2), and a facet's list stops reshuffling under the cursor while you're clicking in it — which is arguably the bigger usability win.
